### DISCUSSION TITLE
post/2020/2020-10-clearing-jupyter-output-p2/

### DISCUSSION THREAD

**Discussion by giscus** [zyrikby/website_related#16] [Announcements]
# post/2020/2020-10-clearing-jupyter-output-p2/

In my previous article, I described why you may need to clear output data in your Jupyter notebooks. As at the time I participated in a pre-sail project for AI Superior, we required a quick solution to achieve this goal. That is why I used Python-based pre-commit framework to create a pipeline to clear output data. However, this approach requires you to install additional Python package into your system, that might not be always possible. Therefore, at the time I decided that I would implement this approach as a pure Bash script. Recently, I have found some spare time and decided to dig deeper into this topic. As a result of my explorations, I developed a git pre-commit hook that clears Jupyter output cells and wrote this article describing it. If you are an adept of ‘show me the code’ and do not want to read the article, you can find the final script here.

https://zhauniarovich.com/post/2020/2020-10-clearing-jupyter-output-p2/

**Comment by davdis** [comment]
I tried the script, but in the case that .ipynb files are modified, the wrong logging is done(while the result is correct, meaning the file is not commited). Could you please point out the error in your implementation?

**Reply by zyrikby** [reply]
You're right. I've modified the script to fix the bug. Thanks!

**Reply by davdis** [reply]
Thanks for taking the time and fixing it!
On my end I also located and fixed the error. I guess the functionality of the script would be the same if you used: 
`if elementIn "$mfile" "${MODIFIED_FILES[@]}"; then`
instead of:
`if elementIn "$mfile" "${IPYNB_FILES[@]}"; then`
Since `MODIFIED_FILES`should be a superset of `IPYNB_FILES`. In most use cases though the two are equivalent.

**Reply by zyrikby** [reply]
In this case, you should iterate over `IPYNB_FILES` in the `for` loop. And in this case, you're right it should be faster.

**Comment by eremingt** [comment]
Thanks for this. Question - why abort the commit if the pre-commit hook modifies ipynb files? Would it be dangerous to have the script re-stage the stripped files and then let the commit proceed?

Also, I want to note a few modifications I made for my environment (MacOS Mojave 10.14.6, PyCharm 2021.2) in case anyone finds this and runs into similar problems:
- PyCharm apparently hates trailing newline at the end of Jupyter notebooks (removing them automatically on every save), even if "Ensure every saved file ends with a line break" is enabled in Editor Preferences. When the pre-commit hook runs, nbconvert adds them back in, which triggers the pre-commit hook to always consider the file modified, and the commit never proceeds. I added `perl -i -0pe 's/\n\Z//' "$FILE"` (based on https://stackoverflow.com/a/16365256/13831553) after the nbconvert command to remove the newline. 
- I think my older version of bash (3.2.57) doesn't like `"${VAR[@]}"` when `VAR` is empty, so I modified three instances to `${VAR[@]+"${VAR[@]}"}` based on https://stackoverflow.com/a/61551944/13831553.

**Reply by zyrikby** [reply]
> Thanks for this. Question - why abort the commit if the pre-commit hook modifies ipynb files? Would it be dangerous to have the script re-stage the stripped files and then let the commit proceed?

As far as I remember, this is the recommendation of this pre-commit - if you modify something during the pre-commit stage, you need to stop commit process and let the developer to check and stage the changes.

**Comment by mingfang** [comment]
This is amazing! Thank you.
